tests/src/python/test_qgssymbollayer.py

```diff
@@ -527,6 +566,23 @@ def testRenderMarkerLayerDataDefined(self):
         symbol.changeSymbolLayer(0, layer)
         points_layer.setRenderer(QgsSingleSymbolRenderer(symbol))
 
+        # Test usedAttributes
+        ctx = QgsRenderContext()
+        exp_ctx = QgsExpressionContext()
+        exp_ctx.appendScope(QgsExpressionContextUtils.globalScope())
+        exp_ctx.appendScope(QgsExpressionContextUtils.projectScope(QgsProject.instance()))
+        exp_ctx.appendScope(points_layer.createExpressionContextScope())
+        ctx.setExpressionContext(exp_ctx)
+        # for symbol layer
+        self.assertEqual(len(layer.usedAttributes(ctx)), 1)
+        self.assertTrue('Class' in layer.usedAttributes(ctx))
+        # for symbol
+        self.assertEqual(len(symbol.usedAttributes(ctx)), 1)
+        self.assertTrue('Class' in symbol.usedAttributes(ctx))
+        # for symbol renderer
+        self.assertEqual(len(points_layer.renderer().usedAttributes(ctx)), 1)
+        self.assertTrue('Class' in points_layer.renderer().usedAttributes(ctx))
```

Comment:
On the same note as @m-kuhn, there's also self.assertIn and self.assertCountEqual (Google it, it doesn't do what you'd expect!) which would be nice here.

Reply:
Thanks to pointing these, I did not know them